tools/gomod-local-update/internal/updater/updater.go

```diff
@@ -0,0 +1,268 @@
+package updater
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"log"
+	"os"
+	"os/exec"
+	"regexp"
+	"strings"
+	"time"
+
+	"golang.org/x/mod/modfile"
+	"golang.org/x/mod/module"
+)
+
+// gitExecutor allows mocking git commands in tests
+type gitExecutor interface {
+	Command(ctx context.Context, args ...string) ([]byte, error)
+}
+
+// realGitExecutor implements actual git command execution
+type realGitExecutor struct{}
+
+func (g *realGitExecutor) Command(ctx context.Context, args ...string) ([]byte, error) {
+	return exec.CommandContext(ctx, "git", args...).Output()
+}
+
+const (
+	goModFile           = "go.mod"
+	goModFileMode       = 0644
+	gitSHALength        = 12
+	gitTimeout          = 30 * time.Second
+	gitTimeFormat       = time.RFC3339
+	gitRemotePattern    = `^[a-zA-Z0-9][-a-zA-Z0-9_.]*$`
+	gitBranchPattern    = `^[a-zA-Z0-9][-a-zA-Z0-9/_]*$`
+	majorVersionPattern = `/v\d+$`
+	shaPattern          = `^[a-fA-F0-9]{40}$` // SHA-1 hashes are 40 hexadecimal characters
+)
+
+type Updater struct {
+	config *Config
+	system SystemOperator
+	git    gitExecutor
+}
+
+// New creates a new Updater
+func New(config *Config, system SystemOperator) *Updater {
+	return &Updater{
+		config: config,
+		system: system,
+		git:    &realGitExecutor{},
+	}
+}
+
+// validateGitInput checks if the remote and branch are in the correct format
+func (u *Updater) validateGitInput(remote, branch string) error {
+	remoteRE := regexp.MustCompile(gitRemotePattern)
+	if !remoteRE.MatchString(remote) {
+		return fmt.Errorf("%w: git remote '%s' contains invalid characters", ErrInvalidConfig, remote)
+	}
+
+	branchRE := regexp.MustCompile(gitBranchPattern)
+	if !branchRE.MatchString(branch) {
+		return fmt.Errorf("%w: git branch '%s' contains invalid characters", ErrInvalidConfig, branch)
+	}
+	return nil
+}
```

Comment:
Would it be better to move this into the `config.go`'s validation?

Seems like this method is only called once with the inputs from the config, so could catch reduce branches by catching it earlier?

Reply:
Makes sense 👍 

Moved in cbda2d1dfeb9d78e12f3dd7391e4e9da25379977.